### PR TITLE
Added a beam optimization example

### PR DESCRIPTION
**Pull request by hwangjt**

### CODE REVIEW

File: openmdao/test_suite/test_examples/beam_optimization/test_beam_optimization.py [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+        prob = Problem(model=BeamGroup(E=E, L=L, b=b, volume=volume, num_elements=num_elements))
+
+        prob.driver = ScipyOptimizer()
+        prob.driver.options['optimizer'] = 'SLSQP'
+        prob.driver.options['tol'] = 1e-9
+        prob.driver.options['disp'] = True
+
+        prob.setup()
+
+        prob.run_model()
```
Comment (Kenneth-T-Moore) [inline openmdao/test_suite/test_examples/beam_optimization/test_beam_optimization.py]: Do we need to run model before running driver?

Comment (hwangjt) [reply openmdao/test_suite/test_examples/beam_optimization/test_beam_optimization.py]: No, I don't believe we do. I removed the `run_model` line

File: openmdao/docs/examples/beam_optimization/beam_optimization.rst [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+    \text{subject to} & & \text{sum}(h) b L_0 = \text{volume} \\
+  \end{array}
+
+where :math:`f` is the vector of forces, :math:`h` is the vector of beam heights, and :math:`L_0` is the length of a single beam element.
+
+The displacements vector :math:`d` is given by
+
+.. math ::
+
+  K d = f ,
```
Comment (swryan) [inline openmdao/docs/examples/beam_optimization/beam_optimization.rst]: extraneous comma?

File: openmdao/test_suite/test_examples/beam_optimization/components/local_stiffness_matrix_comp.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+        self.declare_partials('K_local', 'I',
+            val=self.mtx.reshape(16 * num_elements, num_elements))
+
+    def compute(self, inputs, outputs):
+        num_elements = self.metadata['num_elements']
+        
+        outputs['K_local'] = 0
+        for ind in range(num_elements):
+            outputs['K_local'][ind, :, :] = \
```
Comment (swryan) [inline openmdao/test_suite/test_examples/beam_optimization/components/local_stiffness_matrix_comp.py]: the continuation character doesn't render correctly in the docs

File: openmdao/docs/examples/beam_optimization/beam_optimization.rst [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+===================================================================================
+
+In this example, we optimize the thickness (height) distribution of cantilever beam
+using the adjoint method to compute the gradient.
+We use Euler--Bernoulli beam theory and assume a rectangular section.
+
+Background
+----------
+
+The minimize compliance subject to a constraint on the beam volume.
```
[the remark this answers is not in the document]
Comment (hwangjt) [reply openmdao/docs/examples/beam_optimization/beam_optimization.rst]: Oh, I meant to delete this sentence. Thanks for catching this!

File: openmdao/docs/examples/beam_optimization/beam_optimization.rst (hunk 2 above)
Comment (hwangjt) [reply openmdao/docs/examples/beam_optimization/beam_optimization.rst]: This one was intentional - including a comma here is what's normally done in academic papers

File: openmdao/test_suite/test_examples/beam_optimization/components/local_stiffness_matrix_comp.py (hunk 3 above)
Comment (hwangjt) [reply openmdao/test_suite/test_examples/beam_optimization/components/local_stiffness_matrix_comp.py]: fixed!

**Pull request merged**